docs/developer-guide/code-path-analysis.md

```diff
@@ -0,0 +1,554 @@
+# Code Path Analysis Details
+
+| [Objects](#objects) | [Events](#events) | [Usage Examples](#usage-examples) | [Code Path Examples](#code-path-examples) |
+|:-------------------:|:-----------------:|:---------------------------------:|:-----------------------------------------:|
+
+ESLint's rules can use code paths.
+The code path is execution routes of programs.
+It forks/joins at such as `if` statements.
+
+```js
+if (a && b) {
+    foo();
+}
+bar();
+```
+
+![Code Path Example](./code-path-analysis/helo.svg)
+
+## Objects
+
+Program is expressed with several code paths.
+A code path is expressed with objects of two kinds: `CodePath` and `CodePathSegment`.
+
+### `CodePath`
+
+`CodePath` expresses whole of one code path.
+This object exists for each function and the global.
+This has references of both the initial segment and the final segments of a code path.
+
+`CodePath` has the following properties:
+
+* `id` (`string`) - An unique string. Respective rules can use `id` to save additional information for each code path.
+* `initialSegment` (`CodePathSegment`) - The initial segment of this code path.
+* `finalSegments` (`CodePathSegment[]`) - The final segments which includes both returned and thrown.
+* `returnedSegments` (`CodePathSegment[]`) - The final segments which includes only returned.
+* `thrownSegments` (`CodePathSegment[]`) - The final segments which includes only thrown.
+* `currentSegments` (`CodePathSegment[]`) - Segments of the current position.
+* `upper` (`CodePath|null`) - The code path of the upper function/global scope.
+* `childCodePaths` (`CodePath[]`) - Code paths of functions this code path contains.
+
+### `CodePathSegment`
+
+`CodePathSegment` is a part of a code path.
+A code path is expressed with plural `CodePathSegment` objects, it's similar to doubly linked list.
+Difference from doubly linked list is what there are forking and merging (the next/prev are plural).
+
+`CodePathSegment` has the following properties:
+
+* `id` (`string`) - An unique string. Respective rules can use `id` to save additional information for each segment.
+* `nextSegments` (`CodePathSegment[]`) - The next segments. If forking, there are two or more. If final, there is nothing.
+* `prevSegments` (`CodePathSegment[]`) - The previous segments. If merging, there are two or more. If final, there is nothing.
```

Comment:
I assume "If final, there is nothing." was copied from above? I assume this was supposed to say "If entry, there is nothing" right?